### ISSUE TITLE
Incorporate apt upgrades for security updates

### ISSUE THREAD

**Issue by rarkins** [containerbase/buildpack#139]
Ubuntu base images are only updated periodically, and may have vulnerabilities in packages which are fixed in the apt repository.

How can we balance such upgrading while also retaining *some* caching ability?

Example:

```
ubuntu:20.04 (ubuntu 20.04)
===========================
Total: 2 (UNKNOWN: 0, LOW: 1, MEDIUM: 1, HIGH: 0, CRITICAL: 0)

+-------------+------------------+----------+-------------------+------------------+
|   LIBRARY   | VULNERABILITY ID | SEVERITY | INSTALLED VERSION |  FIXED VERSION   |
+-------------+------------------+----------+-------------------+------------------+
| libgcrypt20 | CVE-2021-40528   | MEDIUM   | 1.8.5-5ubuntu1    | 1.8.5-5ubuntu1.1 |
+             +------------------+----------+                   +                  +
|             | CVE-2021-33560   | LOW      |                   |                  |
+-------------+------------------+----------+-------------------+------------------+
```

Meanwhile the same image with `RUN apt-get update && apt-get -y upgrade && rm -rf /var/lib/apt/lists/*` results in no vulnerabilities.

Ref: https://pythonspeed.com/articles/security-updates-in-docker/

**Comment by viceice** [comment]
😕  Maybe we should create a ubuntu base image, which we rebuild without cache weekly, so we get weekly securitxy updates. Ubuntu base image is updates once a month or so. 🤔

**Comment by wwuck** [comment]
@viceice would this rebuild without cache include running `apt-get -y upgrade`?

Is there any downside to running `apt-get upgrade` when building the renovate image? Docker, OWASP, and hadolint all no longer recommend not upgrading packages in Dockerfiles.

https://github.com/docker/docker.github.io/pull/12571
https://github.com/OWASP/CheatSheetSeries/pull/614
https://github.com/hadolint/hadolint/issues/562

Until renovate gains support for updating from APT repositories (this might not ever be feasible), then we are stuck with running `apt-get upgrade` in our images.  I don't see any easy way to have apt package version pinning in Dockerfiles without a lot of manual effort in tracking updates.  Debian/Ubuntu only keep the latest version in their repositories anyway, so any existing version pins would be quickly broken after a new package version is released.

**Comment by viceice** [comment]
even i add a `apt upgrade` line to the docker file, it would be cached until the parent digest is changed. So it's required to rebuild without cache to force a `apt upgrade` run.

**Comment by viceice** [comment]
we currently don't do any upgrade to suppress too may dupicated files, as docker would save all those new files the a new layer. so worst case it you fully duplicate the base layer

**Comment by rarkins** [comment]
BTW this affects both the base image as well as tools we install such as `curl`

**Comment by viceice** [comment]
yes, we install latest version which is available at the day, when renovate updates the base image or we change some of our base files.

**Comment by rarkins** [comment]
Maybe we could inject a "week number" into our build files so that it invalidates the cached layers weekly? But we might also want the ability to manually trigger it somehow

**Comment by viceice** [comment]
We need to add that value as `fix` commit to the repo so semantic release will do a release and we can use that file as cache buster.

**Comment by wwuck** [comment]
For our images, we are using multi-stage builds with a single RUN command in the intermediate stage to avoid docker caching the `apt-get` commands.

**Comment by rarkins** [comment]
Caching until now has been a desirable feature, not a bug..

**Comment by viceice** [comment]
what about weekly lockfile maintenance as fix release? 😏

**Comment by chrisfarnham** [comment]
@wwuck Could you provide a small example of the multi-stage build with the single RUN that avoides docker caching?

**Comment by wwuck** [comment]
https://github.com/hadolint/hadolint/wiki/DL3009
https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#run

Deleting `rm -rf /var/lib/apt/lists/*` in the same `RUN` statement as `apt-get install` means the package index will always be refreshed on install.  I would be also pinning apt package versions, as [recommended by hadolint](https://github.com/hadolint/hadolint/wiki/DL3008),  if it were possible to automate with renovate.  Unfortunately that is not possible from what I've read in some issues here.

I thought I remember reading somewhere that intermediate stages in a [multi-stage build](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#use-multi-stage-builds) are not cached (I could be wrong here though)?  In any case, I use multi-stage builds to prevent temporary build credentials from appearing in the final image layers.

**Comment by reitzig** [comment]
FWIW, maintaining pinned apt package versions with Renovate is very possible now.

Posting this here since I see Renovate mentioned above (oh hi, @rarkins! :grimacing: ) and we're seeing dozens of mid-level vulnerabilities in renovate/renovate, which is based on this here image (cf. [renovatebot/docker-renovate:Dockerfile](https://github.com/renovatebot/docker-renovate/blob/main/Dockerfile#LL6C6-L6C38)).

**Comment by viceice** [comment]
that would not help, as most vulnerabilities are not solved, so no upgrade at all.
